scipy/optimize/_optimize.py

```diff
@@ -2179,11 +2178,9 @@ def terminate(warnflag, msg):
                     Ap = fhess_p(xk, psupi, *args)
                     hcalls += 1
             else:
-                if isinstance(A, HessianUpdateStrategy):
-                    # if hess was supplied as a HessianUpdateStrategy
-                    Ap = A.dot(psupi)
-                else:
-                    Ap = np.dot(A, psupi)
+                # hess was supplied as a callable or hessian update strategy, so
+                # A is a dense numpy array or sparse matrix
+                Ap = A.dot(psupi)
```

Comment:
I could not notice a performance difference between `np.dot(a, b` and `a.dot(b)` locally. `a.dot(b)` works for sparse and dense arrays so I thought this would simplify things.

Reply:
Didn't using the `@` operator work, as suggested in the linked issue? That'd be more idiomatic and future-proof.

Reply:
As @jeffrey-hokanson pointed out in https://github.com/scipy/scipy/issues/8792#issuecomment-1873466335, [`HessianUpdateStrategy`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.HessianUpdateStrategy.html) does not support `@` yet. That can of course be added but I would prefer to leave that as a follow up, to keep this PR small.

Reply:
That makes sense, thanks for explaining.